### ISSUE TITLE
Transforms some filters into range scans

### ISSUE THREAD

**Issue by evenyag** [GreptimeTeam/greptimedb#2807]
### What type of enhancement is this?

Performance

### What does the enhancement do?

Converts query conditions into primary key ranges to scan.

Here is an example from TiDB's docs: https://docs.pingcap.com/tidb/stable/explain-indexes#indexlookup

```sql
EXPLAIN SELECT * FROM t1 WHERE intkey BETWEEN 300 AND 310;

+-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
| id                            | estRows | task      | access object                  | operator info                     |
+-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
| IndexLookUp_10                | 5.67    | root      |                                |                                   |
| ├─IndexRangeScan_8(Build)     | 5.67    | cop[tikv] | table:t1, index:intkey(intkey) | range:[300,310], keep order:false |
| └─TableRowIDScan_9(Probe)     | 5.67    | cop[tikv] | table:t1                       | keep order:false                  |
+-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
3 rows in set (0.00 sec)
```

### Implementation challenges

We might perform the transformation in the query optimizers.

The storage engine can reduce the key range to scan or rows to return. We can implement range scan in the mito engine.
- The memtable can only scan a subset of keys
- The SST reader can further filter rows by key ranges.

**Comment by KKould** [comment]
Is there a more detailed description or issue to facilitate participation and contribution?
I am very interested in this issue

**Comment by evenyag** [comment]
There are some steps
- implement a detacher to get index ranges to scan from filters. Here is an [example](https://github.com/pingcap/tidb/blob/2c46b8e0041822cb638b97b23a9399b27bc8b6b0/pkg/util/ranger/detacher.go#L863-L879) that we can follow
- support range scan or pruning in both memtables and SSTs
- transform filters to ranges in the query optimizer or the storage engine
- compare with existing min-max and inverted index to see whether we can benefit from it
- merge them into the main branch


We can try to extract ranges from filters in the table [predicate](https://github.com/GreptimeTeam/greptimedb/blob/a6564e72b4022778624d84930fd7350d5c202614/src/table/src/predicate.rs#L55-L58) as the first step. Each range bound could be a `Vec<Value>`.

**Comment by KKould** [comment]
> There are some steps
> 
> * implement a detacher to get index ranges to scan from filters. Here is an [example](https://github.com/pingcap/tidb/blob/2c46b8e0041822cb638b97b23a9399b27bc8b6b0/pkg/util/ranger/detacher.go#L863-L879) that we can follow
> * support range scan or pruning in both memtables and SSTs
> * transform filters to ranges in the query optimizer or the storage engine
> * compare with existing min-max and inverted index to see whether we can benefit from it
> * merge them into the main branch
> 
> We can try to extract ranges from filters in the table [predicate](https://github.com/GreptimeTeam/greptimedb/blob/a6564e72b4022778624d84930fd7350d5c202614/src/table/src/predicate.rs#L55-L58) as the first step. Each range bound could be a `Vec<Value>`.

I thought it might be appropriate that the range bound should be something like a structure like this?
```
struct Range {
   start: Bound<Value>,
   end: Bound<Value>,
}
```

**Comment by evenyag** [comment]
> I thought it might be appropriate that the range bound should be something like a structure like this?
> 
> ```
> struct Range {
>    start: Bound<Value>,
>    end: Bound<Value>,
> }
> ```

A bound might contain multiple columns if we have a compound key. So a bound could be:
```rust
struct Bound {
    values: Vec<Value>,
    inclusive: bool,
}
```
If we encode values by memcomparable encoding, we could also reuse the bound struct here. But I'm not sure whether it is a good idea.
https://github.com/GreptimeTeam/greptimedb/blob/d4a54a085b1197f5e01e696404be331605a46e3d/src/index/src/inverted_index/search/predicate.rs#L40-L47


> I found that `TimeRangePredicateBuilder::extract_time_range_from_expr` is very similar, so can I refer to it for implementation?

Yes. Note that the `TimeRangePredicateBuilder` only supports extracting range for a single column. Extracting ranges is the most complicated thing. I'm also considering the necessity of this issue as we are implementing an [inverted index](https://github.com/GreptimeTeam/greptimedb/issues/2705). We could also construct some skipping indices in memory to speed up queries.

Reference
- https://github.com/pingcap/tidb/blob/60aa838790e36e33b4105d22e191dfa827fe3663/pkg/util/ranger/types.go#L63-L70
- https://github.com/pingcap/tidb/blob/2c46b8e0041822cb638b97b23a9399b27bc8b6b0/pkg/util/ranger/detacher.go#L916-L940

**Comment by KKould** [comment]
Recently I read some articles about index selection, and I found a rule: 
> Use combined indexes when querying with multiple conditions. Note that columns with equivalent conditions need to be placed in the front of the combined index. Here is an example: if the select* from t where c1 = 10 and c2 = 100 and c3 > 10 query is frequently used, consider creating a combined index Index cidx (c1, c2, c3), so that a index prefix can be constructed to scan by query conditions.

So I'm wondering if I can use something like `TimeRangePredicateBuilder` to sequentially obtain the relevant columns in the composite index, and only when the previous one is of equal value can I detach the next column.

Case 1: There is an index with three columns `a, b, c`. When `WHERE a = 1 AND b > 1, the range of a=1 and b >1` can be extracted. The index can use the composite index of `a, b, c`. (if `WHERE a > 1 AND b = 1` then the index cannot be used)

Case 2: There are two indexes with one column `a` and three columns `a, b, c`. When `WHERE (a = 1 AND b > 1) OR a > 1`, the range of a `AND` b cannot be extracted, but at this time the range of a range can be extracted and is `a >= 1`, so you can use a single-column index with column `a`

ref:
- https://docs.pingcap.com/tidb/stable/dev-guide-index-best-practice

**Comment by evenyag** [comment]
> So I'm wondering if I can use something like TimeRangePredicateBuilder to sequentially obtain the relevant columns in the composite index, and only when the previous one is of equal value can I detach the next column.

Yes, it is possible. Note that we should support both `WHERE a = 1 AND b > 1` and `WHERE b > 1 AND a = 1`.

**Comment by KKould** [comment]
> > So I'm wondering if I can use something like TimeRangePredicateBuilder to sequentially obtain the relevant columns in the composite index, and only when the previous one is of equal value can I detach the next column.
> 
> Yes, it is possible. Note that we should support both `WHERE a = 1 AND b > 1` and `WHERE b > 1 AND a = 1`.

Yes, the order should not lead to inconsistent results

I'll try to verify the correctness of this
